File: src/SV5T.Domain/Campaigns/Campaign.cs
using SV5T.Domain.Campaigns.Enums;
using SV5T.Domain.Awards.Enums;
using SV5T.Domain.Standards;
using SV5T.Domain.Submissions;

namespace SV5T.Domain.Campaigns;

public sealed class Campaign : AggregateRoot<Guid>, IAuditableEntity
{
    public Campaign()
    {
        Id = Guid.NewGuid();
    }

    public string Name { get; set; } = string.Empty;
    public string SchoolYear { get; set; } = string.Empty;
    public AwardLevel Level { get; set; }
    public AwardType AwardType { get; set; }
    public CampaignStatus Status { get; set; } = CampaignStatus.Draft;

    // Admin gắn một StandardSet đã Published vào đợt xét.
    public Guid StandardSetId { get; set; }
    public StandardSet StandardSet { get; set; } = null!;

    // City phải trỏ đến School; Central phải trỏ đến City.
    public Guid? PrerequisiteCampaignId { get; set; }
    public Campaign? PrerequisiteCampaign { get; set; }
    public ICollection<Campaign> DependentCampaigns { get; set; } = [];

    // Chỉ dùng cho AwardType.Collective.
    public string CollectiveEligibilityRuleJson { get; set; } = "[]";

    public DateTime RegOpenAt { get; set; }
    public DateTime RegCloseAt { get; set; }
    public DateTime SubmitDeadline { get; set; }
    public DateTime ReviewDeadline { get; set; }
    public string? Description { get; set; }

    public ICollection<Application> Applications { get; set; } = [];

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public string? CreatedBy { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public string? UpdatedBy { get; set; }

    public bool IsRegistrationOpen() =>
        Status == CampaignStatus.Open &&
        DateTime.UtcNow >= RegOpenAt &&
        DateTime.UtcNow <= RegCloseAt;
}

File: frontend/src/types/admin/campaign.ts
/**
 * Campaign Module Types & DTOs
 * Synchronized with backend SV5T.Application.Admin.Dtos & SV5T.Domain.Campaigns
 */

export const AwardLevel = {
  School: 'School',
  City: 'City',
  Central: 'Central',
} as const;
export type AwardLevel = (typeof AwardLevel)[keyof typeof AwardLevel];

export const AwardType = {
  Individual: 'Individual',
  Collective: 'Collective',
} as const;
export type AwardType = (typeof AwardType)[keyof typeof AwardType];

export const CampaignStatus = {
  Draft: 'Draft',
  Open: 'Open',
  Closed: 'Closed',
  Reviewing: 'Reviewing',
  Published: 'Published',
  Archived: 'Archived',
} as const;
export type CampaignStatus = (typeof CampaignStatus)[keyof typeof CampaignStatus];

export const AWARD_LEVEL_LABELS: Record<AwardLevel, string> = {
  [AwardLevel.School]: 'Cấp Trường',
  [AwardLevel.City]: 'Cấp Thành phố',
  [AwardLevel.Central]: 'Cấp Trung ương',
};

export const AWARD_TYPE_LABELS: Record<AwardType, string> = {
  [AwardType.Individual]: 'Cá nhân',
  [AwardType.Collective]: 'Tập thể',
};

export const CAMPAIGN_STATUS_LABELS: Record<CampaignStatus, string> = {
  [CampaignStatus.Draft]: 'Nháp',
  [CampaignStatus.Open]: 'Đang mở đăng ký',
  [CampaignStatus.Closed]: 'Đã đóng đăng ký',
  [CampaignStatus.Reviewing]: 'Đang xét duyệt',
  [CampaignStatus.Published]: 'Đã công bố kết quả',
  [CampaignStatus.Archived]: 'Đã lưu trữ',
};

export interface CampaignResponse {
  id: string;
  name: string;
  schoolYear: string;
  level: AwardLevel;
  awardType: AwardType;
  status: CampaignStatus;
  standardSetId: string;
  standardSetName?: string | null;
  prerequisiteCampaignId?: string | null;
  prerequisiteCampaignName?: string | null;
  regOpenAt: string;
  regCloseAt: string;
  submitDeadline: string;
  reviewDeadline: string;
  description?: string | null;
  createdAt: string;
}

export interface CampaignDetailResponse {
  id: string;
  name: string;
  schoolYear: string;
  level: AwardLevel;
  awardType: AwardType;
  status: CampaignStatus;
  standardSetId: string;
  standardSetName?: string | null;
  prerequisiteCampaignId?: string | null;
  prerequisiteCampaignName?: string | null;
  collectiveEligibilityRuleJson: string;
  regOpenAt: string;
  regCloseAt: string;
  submitDeadline: string;
  reviewDeadline: string;
  description?: string | null;
  totalApplications: number;
  createdAt: string;
}

export interface CreateCampaignRequest {
  name: string;
  schoolYear: string;
  level: AwardLevel;
  awardType: AwardType;
  standardSetId: string;
  prerequisiteCampaignId?: string | null;
  regOpenAt: string;
  regCloseAt: string;
  submitDeadline: string;
  reviewDeadline: string;
  description?: string | null;
  collectiveEligibilityRuleJson: string;
}

export interface UpdateCampaignRequest {
  name: string;
  schoolYear: string;
  level: AwardLevel;
  awardType: AwardType;
  standardSetId: string;
  prerequisiteCampaignId?: string | null;
  regOpenAt: string;
  regCloseAt: string;
  submitDeadline: string;
  reviewDeadline: string;
  description?: string | null;
  collectiveEligibilityRuleJson: string;
}

export interface UpdateCampaignStatusRequest {
  status: CampaignStatus;
}

export interface CampaignFilterParams {
  level?: AwardLevel;
  status?: CampaignStatus;
  schoolYear?: string;
  pageIndex?: number;
  pageSize?: number;
}

export interface PagedResponse<T> {
  items: T[];
  totalCount: number;
  pageIndex: number;
  pageSize: number;
  totalPages: number;
}
